### ISSUE TITLE
Error generating SDK

### ISSUE THREAD

**Issue by R2D221** [CenterEdge/Yardarm#239]
I'm trying to generate an API, specifically for this: https://www.eia.gov/opendata/eia-api-swagger.zip

First I tried using an MSBuild project, but the error shown isn't very helpful:

```
1>C:\Users\atorres\.nuget\packages\yardarm.sdk\0.4.1\Sdk\Sdk.targets(148,5): error : Error generating SDK
1>Done building project "EiaApi.csproj" -- FAILED.
```

So I installed the command line instead, and it apparently breaks in two different parts. Here's the output:

```
[INF] Restoring packages for EiaApi...
[INF] Restoring packages for .NETCoreApp,Version=v6.0...
[INF] Resolving conflicts for net6.0...
[INF] Checking compatibility of packages on net6.0.
[INF] All packages and projects are compatible with net6.0.
[WRN] Unsupported security scheme 'api_key', implementing no-op placeholder.
[ERR] Error generating SDK
System.AggregateException: One or more errors occurred. (Object reference not set to an instance of an object.) (The method or operation is not implemented.)
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.SyntaxToken.SyntaxIdentifier..ctor(String text)
   at Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.SyntaxToken.SyntaxIdentifierExtended..ctor(SyntaxKind contextualKind, String text, String valueText)
   at Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.SyntaxToken.SyntaxIdentifierWithTrivia..ctor(SyntaxKind contextualKind, String text, String valueText, GreenNode leading, GreenNode trailing)
   at Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.SyntaxToken.Identifier(GreenNode leading, String text, GreenNode trailing)
   at Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.SyntaxToken.Identifier(SyntaxKind contextualKind, GreenNode leading, String text, String valueText, GreenNode trailing)
   at Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.SyntaxFactory.Identifier(SyntaxKind contextualKind, GreenNode leading, String text, String valueText, GreenNode trailing)
   at Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.SyntaxFactory.Identifier(GreenNode leading, String text, GreenNode trailing)
   at Microsoft.CodeAnalysis.CSharp.SyntaxFactory.Identifier(String text)
   at Microsoft.CodeAnalysis.CSharp.SyntaxFactory.MethodDeclaration(TypeSyntax returnType, String identifier)
   at Yardarm.Generation.Tag.TagTypeGeneratorBase.GenerateOperationMethodHeader(ILocatedOpenApiElement`1 operation)+MoveNext() in /home/runner/work/Yardarm/Yardarm/src/main/Yardarm/Generation/Tag/TagTypeGeneratorBase.cs:line 26
   at System.Linq.Enumerable.SelectManyIterator[TSource,TCollection,TResult](IEnumerable`1 source, Func`2 collectionSelector, Func`3 resultSelector)+MoveNext()
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at Microsoft.CodeAnalysis.SyntaxList`1.CreateNode(IEnumerable`1 nodes)
   at Microsoft.CodeAnalysis.SyntaxList`1..ctor(IEnumerable`1 nodes)
   at Microsoft.CodeAnalysis.CSharp.SyntaxFactory.List[TNode](IEnumerable`1 nodes)
   at Yardarm.Generation.Tag.TagTypeGenerator.GenerateInterface() in /home/runner/work/Yardarm/Yardarm/src/main/Yardarm/Generation/Tag/TagTypeGenerator.cs:line 40
   at Yardarm.Generation.Tag.TagTypeGenerator.Generate()+MoveNext() in /home/runner/work/Yardarm/Yardarm/src/main/Yardarm/Generation/Tag/TagTypeGenerator.cs:line 35
   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at Yardarm.Generation.TypeGeneratorBase.GenerateSyntaxTree() in /home/runner/work/Yardarm/Yardarm/src/main/Yardarm/Generation/TypeGeneratorBase.cs:line 37
   at Yardarm.Generation.Tag.TagGenerator.Generate()+MoveNext() in /home/runner/work/Yardarm/Yardarm/src/main/Yardarm/Generation/Tag/TagGenerator.cs:line 32
   at System.Linq.Parallel.SelectManyQueryOperator`3.SelectManyQueryOperatorEnumerator`1.MoveNext(TOutput& currentElement, Pair`2& currentKey)
   at System.Linq.Parallel.StopAndGoSpoolingTask`2.SpoolingWork()
   at System.Linq.Parallel.SpoolingTaskBase.Work()
   at System.Linq.Parallel.QueryTask.BaseWork(Object unused)
   at System.Linq.Parallel.QueryTask.<>c.<.cctor>b__10_0(Object o)
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
   --- End of inner exception stack trace ---
   at System.Linq.Parallel.QueryTaskGroupState.QueryEnd(Boolean userInitiatedDispose)
   at System.Linq.Parallel.SpoolingTask.SpoolStopAndGo[TInputOutput,TIgnoreKey](QueryTaskGroupState groupState, PartitionedStream`2 partitions, SynchronousChannel`1[] channels, TaskScheduler taskScheduler)
   at System.Linq.Parallel.DefaultMergeHelper`2.System.Linq.Parallel.IMergeHelper<TInputOutput>.Execute()
   at System.Linq.Parallel.MergeExecutor`1.Execute()
   at System.Linq.Parallel.MergeExecutor`1.Execute[TKey](PartitionedStream`2 partitions, Boolean ignoreOutput, ParallelMergeOptions options, TaskScheduler taskScheduler, Boolean isOrdered, CancellationState cancellationState, Int32 queryId)
   at System.Linq.Parallel.PartitionedStreamMerger`1.Receive[TKey](PartitionedStream`2 partitionedStream)
   at System.Linq.Parallel.SelectManyQueryOperator`3.WrapPartitionedStreamNotIndexed[TLeftKey](PartitionedStream`2 inputStream, IPartitionedStreamRecipient`1 recipient, QuerySettings settings)
   at System.Linq.Parallel.SelectManyQueryOperator`3.WrapPartitionedStream[TLeftKey](PartitionedStream`2 inputStream, IPartitionedStreamRecipient`1 recipient, Boolean preferStriping, QuerySettings settings)
   at System.Linq.Parallel.UnaryQueryOperator`2.UnaryQueryOperatorResults.ChildResultsRecipient.Receive[TKey](PartitionedStream`2 inputStream)
   at System.Linq.Parallel.ListQueryResults`1.GivePartitionedStream(IPartitionedStreamRecipient`1 recipient)
   at System.Linq.Parallel.UnaryQueryOperator`2.UnaryQueryOperatorResults.GivePartitionedStream(IPartitionedStreamRecipient`1 recipient)
   at System.Linq.Parallel.QueryOperator`1.ExecuteAndGetResultsAsArray()
   at System.Linq.ParallelEnumerable.ToArray[TSource](ParallelQuery`1 source)
   at Yardarm.Enrichment.Compilation.SyntaxTreeCompilationEnricher.EnrichAsync(CSharpCompilation target, CancellationToken cancellationToken) in /home/runner/work/Yardarm/Yardarm/src/main/Yardarm/Enrichment/Compilation/SyntaxTreeCompilationEnricher.cs:line 27
   at Yardarm.Enrichment.EnrichmentExtensions.<>c__3`1.<EnrichAsync>b__3_0(TTarget p, IAsyncEnricher`1 enricher, CancellationToken ct) in /home/runner/work/Yardarm/Yardarm/src/main/Yardarm/Enrichment/EnrichmentExtensions.cs:line 31
   at System.Linq.AsyncEnumerable.<AggregateAwaitWithCancellationAsyncCore>g__Core|7_0[TSource,TAccumulate](IAsyncEnumerable`1 source, TAccumulate seed, Func`4 accumulator, CancellationToken cancellationToken) in /_/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/Aggregate.cs:line 214
   at System.Linq.AsyncEnumerable.<AggregateAwaitWithCancellationAsyncCore>g__Core|7_0[TSource,TAccumulate](IAsyncEnumerable`1 source, TAccumulate seed, Func`4 accumulator, CancellationToken cancellationToken) in /_/Ix.NET/Source/System.Linq.Async/System/Linq/Operators/Aggregate.cs:line 212
   at Yardarm.YardarmGenerator.BuildForTargetFrameworkAsync(GenerationContext context, NuGetFramework targetFramework, Stream dllOutput, Stream pdbOutput, Stream xmlDocumentationOutput, Stream referenceAssemblyOutput, CancellationToken cancellationToken) in /home/runner/work/Yardarm/Yardarm/src/main/Yardarm/YardarmGenerator.cs:line 128
   at Yardarm.YardarmGenerator.EmitAsync(CancellationToken cancellationToken) in /home/runner/work/Yardarm/Yardarm/src/main/Yardarm/YardarmGenerator.cs:line 56
   at Yardarm.CommandLine.GenerateCommand.ExecuteAsync(CancellationToken cancellationToken) in /home/runner/work/Yardarm/Yardarm/src/main/Yardarm.CommandLine/GenerateCommand.cs:line 73
 ---> (Inner Exception #1) System.NotImplementedException: The method or operation is not implemented.
   at Yardarm.Generation.Request.ParameterTypeGenerator.GetTypeInfo() in /home/runner/work/Yardarm/Yardarm/src/main/Yardarm/Generation/Request/ParameterTypeGenerator.cs:line 18
   at Yardarm.Generation.TypeGeneratorBase`1.GetSourceFilePath() in /home/runner/work/Yardarm/Yardarm/src/main/Yardarm/Generation/TypeGeneratorBase`1.cs:line 40
   at Yardarm.Generation.TypeGeneratorBase.GenerateSyntaxTree() in /home/runner/work/Yardarm/Yardarm/src/main/Yardarm/Generation/TypeGeneratorBase.cs:line 43
   at Yardarm.Generation.Request.ParameterGenerator.Generate(ILocatedOpenApiElement`1 parameter) in /home/runner/work/Yardarm/Yardarm/src/main/Yardarm/Generation/Request/ParameterGenerator.cs:line 33
   at System.Linq.Utilities.<>c__DisplayClass2_0`3.<CombineSelectors>b__0(TSource x)
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
   at Yardarm.Generation.Request.ParameterGenerator.Generate()+MoveNext() in /home/runner/work/Yardarm/Yardarm/src/main/Yardarm/Generation/Request/ParameterGenerator.cs:line 23
   at System.Linq.Parallel.SelectManyQueryOperator`3.SelectManyQueryOperatorEnumerator`1.MoveNext(TOutput& currentElement, Pair`2& currentKey)
   at System.Linq.Parallel.StopAndGoSpoolingTask`2.SpoolingWork()
   at System.Linq.Parallel.SpoolingTaskBase.Work()
   at System.Linq.Parallel.QueryTask.BaseWork(Object unused)
   at System.Linq.Parallel.QueryTask.<>c.<.cctor>b__10_0(Object o)
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__272_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)<---
```

What I found interesting is that apparently it breaks on a not implemented method right here:

https://github.com/CenterEdge/Yardarm/blob/23681d1e29e68e78a81a952245a7c5d10401e1ba/src/main/Yardarm/Generation/Request/ParameterTypeGenerator.cs#L18

So I don't know if what I'm doing is not supported or if I'm doing something wrong myself.

**Comment by brantburnett** [comment]
Interesting. For a new project, I'd recommend trying the 0.5.0 beta (it'll probably be GA soon). This will avoid some various breaking changes moving from 0.4 to 0.5 and includes several improvements.

Do you still get the same error using [0.5.0-beta0002](https://github.com/CenterEdge/Yardarm/releases/tag/release%2F0.5.0-beta0002)?

**Comment by R2D221** [comment]
I updated to version 0.5.0-beta0002 and I still get the error.

I also forgot. The command I'm running is this:

    yardarm generate -n "EiaApi" -i eia-api-swagger.yaml --embed -f "net6.0"

**Comment by brantburnett** [comment]
It appears that this particular specification uses a feature that hasn't been implemented in Yardarm yet, a shared request body definition in `components.requestBodies`. I'm going to look into what it will take to implement this feature.

**Comment by brantburnett** [comment]
Actually, there are several other issues with compatibility. I'm trying to work through them all, though it may take me a couple days.

One of the issues is that some of the operations are missing `operationId` values to give the operation a name, which is used to generate class and method names in C#. The simplest solution is to skip those operations and only provide operations with ids. Would you like to use any of these operations that are missing ids? Or do you have the option to fill in that attribute before generation? If not, I'll need to work on a deterministic id generation process.

**Comment by brantburnett** [comment]
@R2D221 

At this point, I believe I have all of the issues I can see resolved other than operations missing ids. For those, we have three options:

1. Fixup the specification file to have operation IDs before you generate
2. Implement an extension that registers a custom `IOperationNameProvider` which provides names for those operations
3. Come up with some deterministic approach that is broadly acceptable which we can apply to the built-in `DefaultOperationNameProvider`

Let me know your thoughts. I plan to wait to make another 0.5.0 beta since we may need to choose option 3.

**Comment by brantburnett** [comment]
I released `0.5.0-beta0003` with all the fixes and improvements to support the referenced spec. I will close this issue for now, but feel free to reopen it if you need further assistance.

**Comment by R2D221** [comment]
Sorry, I wasn't getting notificacions until now. Yes, let me test it and I'll get back to you
